File: model/menu/menu.go
package menu

import (
	"github.com/ipp-ementa/iped/model/customerror"
	"github.com/ipp-ementa/iped/model/dish"
	"github.com/jinzhu/gorm"
)

// Menu is a model that contains a set of dishes available at either lunch or dinner
// A UML overview of this model can be found at https://github.com/ipp-ementa/iped-documentation/wiki/Architecture#models-structure
type Menu struct {
	gorm.Model
	MenuEntryID uint `gorm:"type:int REFERENCES menu_entries(id) ON UPDATE CASCADE ON DELETE CASCADE"`
	Type        MenuType
	DishesSlice []dish.Dish
}

// New initializes a Menu model using a menu type and a set of dishes
// A FieldError is returned either if the menu type is invalid, no dishes were provided or the dishes are not unique
func New(Type int, Dishes []dish.Dish) (Menu, *customerror.FieldError) {

	menu := Menu{gorm.Model{}, 0, MenuType(Type), Dishes}

	err := grantValidMenuType(Type)

	if err != nil {
		return menu, err
	}

	err = grantThatAtLeastOneDishWasProvided(Dishes)

	if err != nil {
		return menu, err
	}

	err = grantNoDuplicatedDishesExist(Dishes)

	return menu, err
}

// Dishes returns the available dishes on a menu as a slice
// The returned slice has different reference of the one in Menu struct
// In order to prevent modifications
func (menu Menu) Dishes() []dish.Dish {
	availableDishes := make([]dish.Dish, len(menu.DishesSlice))

	copy(availableDishes, menu.DishesSlice)

	return availableDishes
}

// This function grants that a menu type is valid, and if not returns a FieldError
// See [MenuType.Validate] for validation logic
func grantValidMenuType(menutype int) *customerror.FieldError {

	var err *customerror.FieldError

	if !Validate(menutype) {
		err = &customerror.FieldError{Field: "menutype", Model: "menu", Explanation: "specified menu type is not valid"}
	}

	return err
}

// This function grants that at least one dish is provided in given dish slice
// If the given dish slice is nil or empty a FieldError is returned
func grantThatAtLeastOneDishWasProvided(dishes []dish.Dish) *customerror.FieldError {

	var err *customerror.FieldError

	if dishes == nil || len(dishes) == 0 {
		err = &customerror.FieldError{Field: "dishes", Model: "menu", Explanation: "menu requires at least one dish to be provided"}
	}

	return err
}

// This function grants that all dishes given in a slice are unique
// If a dish proves equality to any other dish in the slice, a FieldError is returned
func grantNoDuplicatedDishesExist(dishes []dish.Dish) *customerror.FieldError {

	var err *customerror.FieldError

	unique := true
	dishesLength := len(dishes)
	i := 0

	for i < dishesLength {
		j := i + 1
		for j < dishesLength {
			unique = !dishes[i].Equals(dishes[j])
			j++
			if !unique {
				i = dishesLength
				j = i
			}
		}
		i++
	}

	if !unique {
		err = &customerror.FieldError{Field: "dishes", Model: "menu", Explanation: "menu cannot have duplicated dishes"}
	}

	return err
}

File: models/menu.ts
import { Dish } from "./dish.ts";
import { HasDuplicates, Ok, Result, Err } from "./common.ts";
import { v4 } from "../deps.ts";

class Menu {
  id: string;

  dishes: Dish[];

  type: MenuType;

  private constructor(dishes: Dish[], type: MenuType, id: string) {
    this.dishes = dishes;
    this.type = type;
    this.id = id;
  }

  public static create(dishes: Dish[], type: MenuType): Result<Menu, string> {
    if (dishes.length == 0) {
      return Err("must provide at least one dish");
    } else if (HasDuplicates<Dish>(dishes)) {
      return Err("cannot have duplicate dish");
    } else {
      return Ok(new Menu(dishes, type, `${MenuType[type]}_${v4.generate()}`));
    }
  }

  public static fromJson(object: any): Menu {
    return new Menu(
      object.dishes,
      object.type,
      object.id,
    );
  }

  public static typeFromString(type: string): Result<MenuType, string> {
    switch (type) {
      case MenuType[0]:
        return Ok(MenuType.lunch);
      case MenuType[1]:
        return Ok(MenuType.dinner);
      default:
        return Err(`type: ${type} is not valid`);
    }
  }
}

enum MenuType {
  lunch,
  dinner,
}

export { Menu, MenuType };
